### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Jakeii**
This initially started with just wanting to have sanitisers support nested params, but I found the existing code quite hard to follow, and doing a few odd things. So I ended up changing the whole thing, but all the existing tests pass with some new ones.

The result of all this is... sanitisers support nested params, and are chainable. Added some documentation to the code describing whats going on. And added documentation in README.md

Only one breaking change, in order make sanitisers chainable, to do that I couldn't have the sanitizers return the result directly, to do that now you need to append `.result()`.

### CODE REVIEW

**Comment by rustybailey** [comment]
This is a lot to digest, so I will need to look over this tonight. Also, it looks like this should solve https://github.com/ctavan/express-validator/issues/58.

Since it includes a breaking change, we may need to bump to 3.0.0 (in which case, I may bring in some easy changes I've been putting off, but I'll see how easy they are).

**Comment by Jakeii** [comment]
No problem! I also had another feature to add that allow validating all properties using a schema a la [ember-validations](https://github.com/dockyard/ember-validations) I'll put that in a separate pull request when it's finished.

File: README.md [hunk 1]
```diff
@@ -44,9 +53,9 @@ app.post('/:urlparam', function(req, res) {
     return;
   }
   res.json({
-    urlparam: req.param('urlparam'),
-    getparam: req.param('getparam'),
-    postparam: req.param('postparam')
+    urlparam: req.params.urlparam,
+    getparam: req.params.getparam,
+    postparam: req.param.postparam
```
Comment (rustybailey) [inline README.md]: Typo - should be "params"

File: README.md [hunk 2]
```diff
@@ -195,40 +232,51 @@ req.checkBody('email').optional().isEmail();
...
+console.log(req.body.comment.user); // 'a user'
 
-```javascript
-req.assert(['user', 'fields', 'email'], 'valid email required').isEmail();
-var errors = req.validationErrors();
-console.log(errors);
 ```
 
-Output:
+Sanitizes the specified parameter (using 'dot-notaion' or array), the parameter will be updated to the sanitized result. Can be chained like validators. See [chriso/validator.js](https://github.com/chriso/validator.js) for available sanitizers, or [add your own](#customsanitizers).
```
Comment (rustybailey) [inline README.md]: Typo - "notaion" -> "notation"

File: lib/express_validator.js [hunk 3]
```diff
@@ -1,262 +1,335 @@
...
- * names rather than the actual strings.
- *
- *
- * 1. To validate parameters, use `req.check(param_name, [err_message])`
- *        e.g. req.check('param1').len(1, 6).isInt();
- *        e.g. req.checkHeader('referer').contains('mydomain.com');
+var validator = require('validator');
+
+// validators and sanitizers not prefixed with is/to
+var additonalValidators = ['contains', 'equals', 'matches'];
```
Comment (rustybailey) [inline lib/express_validator.js]: "additional" is misspelled in `additonalValidators` and `additonalSanitizers`. Be sure to correct this variable in all places it occurs.

File: lib/express_validator.js [hunk 4]
```diff
@@ -1,262 +1,335 @@
...
       };
+    });
+
+    req.checkFiles = function(param, failMsg) {
+      return new ValidatorChain(param, failMsg, req, 'files', options);
+    };
 
-      var methods = [];
+    req.checkHeaders = function(param, failMsg) {
+      if(param === 'referrer') {param = 'referer';}
```
Comment (rustybailey) [inline lib/express_validator.js]: I'm just going to comment on this first instance, but throughout the refactor you have a lot of one-line conditionals. I realize now that I should have put a code style guide in this repo, but I would prefer:
1) A space between the "if" and opening parens
2) A space between the closing parens and the opening bracket
3) Conditionals to be multi-line (unless they are ternary operators)

File: lib/express_validator.js [hunk 5]
```diff
@@ -1,262 +1,335 @@
...
+function makeValidator(methodName, container) {
+  return function() {
+    var args = [];
+    args.push(this.value);
+    args = args.concat(Array.prototype.slice.call(arguments));
 
-      methods['len'] = function() {
-        return methods.isLength.apply(methods.isLength, Array.prototype.slice.call(arguments));
-      };
+    var isValid =  (this.skipValidating) ? true : container[methodName].apply(container, args);
```
Comment (rustybailey) [inline lib/express_validator.js]: There are two space in front of this equals sign. Also, it might be more readable to assign `container[methodName].apply(container, args)` to a variable and then just have `var isValid =  this.skipValidating || newVariable;`

File: lib/express_validator.js [hunk 6]
```diff
@@ -1,262 +1,335 @@
...
+ * @param  {(string|string[])} name [description]
+ * @return {string}
+ */
 
-    req.filter = function(item) {
-      return sanitize(this, item, param(req, item));
-    };
+function locate(req, name){
+  if(get(req.params, name)) {
+    return  'params';
```
Comment (rustybailey) [inline lib/express_validator.js]: Two spaces between return and 'params'

File: lib/express_validator.js [hunk 7]
```diff
@@ -1,262 +1,335 @@
...
+        obj[key] = defaults[key];
       }
-      return defaultValue;
     }
+  }
+  return obj;
+}
+
+function capitalCase(string) {
+    return string.charAt(0).toUpperCase() + string.slice(1);
```
Comment (rustybailey) [inline lib/express_validator.js]: 4 spaces here when there should be 2.

File: lib/express_validator.js [hunk 8]
```diff
@@ -1,262 +1,335 @@
...
+ * @param  {string}             failMsg       validation failure message
+ * @param  {Request}            req           request to attach validation errors
+ * @param  {string}             location      request property to find value (body, params, query, etc.)
+ * @param  {object}             options       options containing error formatter
+ */
+
+var ValidatorChain = function(param, failMsg, req, location, options) {
+  this.errorFormatter = options.errorFormatter;
+  this.param = param;
+  this.value = (location) ? get(req[location], param) : undefined;
```
Comment (rustybailey) [inline lib/express_validator.js]: Not really necessary to have parens around `location`

File: lib/express_validator.js [hunk 9]
```diff
@@ -1,262 +1,335 @@
...
+        return errors;
       }
-    });
-    Object.keys(_options.customSanitizers).forEach(function(customName) {
-      methods[customName] = function() {
-        var args = [value].concat(Array.prototype.slice.call(arguments));
-        var result = _options.customSanitizers[customName].apply(null, args);
-        request.updateParam(param, result);
-        return result;
+      return (req._validationErrors.length > 0) ? req._validationErrors : false;
```
Comment (rustybailey) [inline lib/express_validator.js]: Not really necessary to have parens around `req._validationErrors.length > 0`

File: lib/express_validator.js [hunk 10]
```diff
@@ -1,262 +1,335 @@
...
-      // For regex matches you can access the parameters using numbers.
-      if (!Array.isArray(param)) {
-        param = typeof param === 'number' ?
-          [param] :
-          param.split('.').filter(function(e){
-            return e !== '';
-          });
-      }
+    req.sanitize = function(param) {
+      return new SanitizerChain(param, req, ['params', 'query', 'body']);
```
Comment (rustybailey) [inline lib/express_validator.js]: Can we reuse the `locations` variable here? Should locations be `['params', 'query', 'body']` or `['body', 'params', 'query']`?

**Comment by rustybailey** [comment]
I finally had time to review this. I can confirm the tests still run locally. Most of my comments were syntax-related, but please let me know if you have any questions.

Seeing as the `result()` method is the only breaking change, is there any way you can think of to return the sanitizer's result without explicitly calling a method? Just to confirm my understanding, the sanitized value is still set on the req object without calling `result()`. Calling `result()` is solely for assigning the resulting sanitized value to a new variable, correct?

If you have a chance, @ctavan, I would love for you to glance over this. Since you are more intimately familiar with this codebase, it would be good to have another set of eyes on a refactor of this size.

File: lib/express_validator.js [hunk 11]
```diff
@@ -1,262 +1,335 @@
...
- *    where errback receives a parameter containing the error message
+ * @class
+ * @param  {(string|string[])}  param         path to property to validate
+ * @param  {string}             failMsg       validation failure message
+ * @param  {Request}            req           request to attach validation errors
+ * @param  {string}             location      request property to find value (body, params, query, etc.)
+ * @param  {object}             options       options containing error formatter
+ */
+
+var ValidatorChain = function(param, failMsg, req, location, options) {
```
Comment (ctavan) [inline lib/express_validator.js]: For "constructors" that will be used with `new` I'd prefer the notation:

```js
function ValidatorChain(param, …) {
}

ValidatorChain.prototype.result = function() {
};
```

File: lib/express_validator.js [hunk 12]
```diff
@@ -1,262 +1,335 @@
...
+    }
+  };
 
-  var _options = {};
+  options = extend(defaults, options);
 
-  _options.customValidators = options.customValidators || {};
-  _options.customSanitizers = options.customSanitizers || {};
+  // set validators and sainitzers as prototype methods on corresponding chains
+  Object.keys(validator).forEach(function(methodName) {
```
Comment (ctavan) [inline lib/express_validator.js]: Might be a good point to introduce the underscore library to be able to write:

_.each(validator, function(validatorMethod, methodName

File: lib/express_validator.js [hunk 13]
```diff
@@ -1,262 +1,335 @@
...
+    if(methodName.match(/^is/) || additonalValidators.indexOf(methodName) > -1) {
```
Comment (ctavan) [inline lib/express_validator.js]: Here you could use `_.contains(additonalValidators, methodName)` if you had underscore...

File: lib/express_validator.js (hunk 10 above)
Comment (ctavan) [inline lib/express_validator.js]: :+1:

File: lib/express_validator.js (hunk 4 above)
Comment (ctavan) [inline lib/express_validator.js]: :+1:

Maybe time to introduce jshint and jscs? ;)

File: lib/express_validator.js [hunk 14]
```diff
@@ -1,262 +1,335 @@
...
-        return req.params[name];
-      }
-      // query string params
-      if (undefined !== req.query[name]) {
-        return req.query[name];
-      }
-      // request body params via connect.bodyParser
-      if (req.body && undefined !== req.body[name]) {
-        return req.body[name];
+function extend(defaults, options){
```
Comment (ctavan) [inline lib/express_validator.js]: Please consider using underscore.js. You can then use `_.extend()`.

**Comment by ctavan** [comment]
Great work! I did a quick review (not to in-depth) and all in all it looks really good! My main suggestion would be to introduce underscore.js as a dependency since it will allow you to shorten the code and make it even more readable in quite some places.

**Comment by rustybailey** [comment]
Thanks for looking over this @ctavan! @Jakeii, please let us know if you have any questions while going through these comments.

File: lib/express_validator.js (hunk 4 above)
Comment (rustybailey) [inline lib/express_validator.js]: We already have a `.jshintrc`, but at the moment we're relying on each contributor to be using jshint locally and we're not programmatically enforcing it. After this gets merged in, I may add that command as a check in Travis. Same with jscs.

**Comment by Jakeii** [comment]
Sweet, thanks @ctavan, @rustybailey! I'll go over all the suggestions, and update soon.

File: lib/express_validator.js [hunk 15]
```diff
@@ -1,262 +1,335 @@
...
-        return this.body[name] = value;
-      }
-      return false;
-    };
+    }.bind(this));
 
-    req.check = checkParam(req, function(item) {
-      return param(req, item);
-    });
+    return this;
```
Comment (Jakeii) [inline lib/express_validator.js]: @rustybailey RE the breaking change, sure chaining sanitizers could be exchanged for compatibility by returning `value` rather that `this` here.

Comment (rustybailey) [inline lib/express_validator.js]: If this is the only thing, yeah, let's change it back to `value` just to make this a frictionless update. If someone specifically needs multiple sanitizers on one value, I would suggest they use multiple lines or create a custom sanitizer. Obviously, update docs/unit tests related to this.

**Comment by Jakeii** [comment]
Ok the typos have been fixed, and suggestions incorporated! Should have full compatibility now.

**Comment by ctavan** [comment]
Great @Jakeii ! Would you mind squashing all your commits together so we get a clean merge without any states that break the tests?

**Comment by rustybailey** [comment]
Thanks, @Jakeii! Just had a few more small comments. Let me know when you have completed those and @ctavan's request, and I can have this merged in. I also want to merge in https://github.com/ctavan/express-validator/pull/141 after yours tonight and then I should be able to bump minor version and deploy to npm.

**Comment by Jakeii** [comment]
Should be all neat and tidy now, corrected those last few typos.

**Comment by rustybailey** [comment]
I noticed that the `result()` method and the associated tests are still present. Is this method still needed? I thought were were changing the sanitizer to return `value` instead of `this`.

**Comment by rustybailey** [comment]
@Jakeii In that same vein, it is still named SanitizerChain even though the docs advertise it as not being chainable. Just need some clarification here.

**Comment by Jakeii** [comment]
@rustybailey my bad I've renamed SanitizerChain  to Sanitizer, removed `result()`, return the result in sanitizers, and fixed the tests.
